### PR TITLE
balance hot regions by peer && by leader

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
We just consider the leader when do "hot regions" balance,  this will lead to a bad situation that a store contains many hot regions’ peer, and TiKV nodes load various from each other.
@nolouch @disksing @siddontang PTAL

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -398,25 +404,33 @@ func (m RegionsStat) Less(i, j int) bool { return m[i].WrittenBytes < m[j].Writt
...
 type StoreHotRegions struct {
-	TotalWrittenBytes uint64      `json:"total_written"`
-	RegionCount       int         `json:"region_count"`
-	RegionsStat       RegionsStat `json:"stats"`
+	// Hot regions in this store as the role of Leader
+	WrittenBytesAsLeader uint64      `json:"total_written_bytes_as_leader"`
+	RegionsCountAsLeader int         `json:"regions_count_as_leader"`
+	RegionsStatAsLeader  RegionsStat
+
+	// Hot regions in this store as the role of Follower or Follower
```
Comment (siddontang) [inline server/balancer.go]: what is `of Follower or Follower`?

File: server/balancer.go [hunk 2]
```diff
@@ -468,183 +475,249 @@ func (h *balanceHotRegionScheduler) calculateScore(cluster *clusterInfo) {
...
 	var hotRegionTotalCount float64
 	for _, m := range h.scoreStatus {
-		hotRegionTotalCount += float64(m.RegionsStat.Len())
+		hotRegionTotalCount += float64(m.RegionsStatAsPeer.Len())
 	}
 
 	avgRegionCount := hotRegionTotalCount / float64(len(h.scoreStatus))
 	// Multiplied by hotRegionLimitFactor to avoid transfer back and forth
-	limit := uint64((float64(s.RegionsStat.Len()) - avgRegionCount) * hotRegionLimitFactor)
+	limit := uint64((float64(s.RegionsStatAsPeer.Len()) - avgRegionCount) * hotRegionLimitFactor)
```
Comment (nolouch) [inline server/balancer.go]: need we adjust limit by leader?

**Comment by siddontang** [comment]
PTAL @disksing @nolouch

File: server/balancer.go [hunk 3]
```diff
@@ -468,183 +475,268 @@ func (h *balanceHotRegionScheduler) calculateScore(cluster *clusterInfo) {
...
+
+type StoreHotRegionInfos struct {
+	AsPeer map[uint64]*HotRegionsStat `json:"as_peer"`
+	AsLeader map[uint64]*HotRegionsStat `json:"as_leader"`
+}
+
+func (h *balanceHotRegionScheduler) GetStatus() *StoreHotRegionInfos {
+	h.RLock()
+	defer h.RUnlock()
+	asPeer := make(map[uint64]*HotRegionsStat)
```
Comment (siddontang) [inline server/balancer.go]: make with capacity

File: server/balancer.go [hunk 4]
```diff
@@ -468,183 +475,268 @@ func (h *balanceHotRegionScheduler) calculateScore(cluster *clusterInfo) {
...
+			srcStoreID = storeID
 			continue
 		}
 
-		if maxHotStoreRegionCount == s.RegionsStat.Len() && maxWritten < s.TotalWrittenBytes {
-			maxWritten = s.TotalWrittenBytes
-			sourceStore = sid
+		if maxHotStoreRegionCount == statistics.RegionsStat.Len() && maxWrittenBytes < statistics.WrittenBytes {
+			maxWrittenBytes = statistics.WrittenBytes
+			srcStoreID = storeID
```
Comment (disksing) [inline server/balancer.go]: I suppose you can make it simpler:
```
count, writtenBytes := statistics.RegionStat.Len(), statistics.WrittenBytes
if count >= 2 && (count > maxRegionCount || count == maxRegionCount && writtenBytes > maxWrittenBytes) {
    maxRegionCount, maxWrittenBytes = count, writtenBytes
    srcStoreID = storeID
}
```

File: server/balancer.go [hunk 5]
```diff
@@ -468,183 +475,268 @@ func (h *balanceHotRegionScheduler) calculateScore(cluster *clusterInfo) {
...
+			if filterTarget(store, filters) {
+				continue
+			}
+			destStoreIDs = append(destStoreIDs, store.GetId())
+		}
+
+		destStoreID = h.selectDestStoreByPeer(destStoreIDs, srcRegion, srcStoreID)
+		if destStoreID != 0 {
+			srcRegion.WrittenBytes = rs.WrittenBytes
+			h.adjustBalanceLimitByPeer(srcStoreID)
```
Comment (nolouch) [inline server/balancer.go]: If the peer is relatively balanced, the limit will not adjust when balance leader.

**Comment by siddontang** [comment]
Any update?

**Comment by zhangjinpeng1987** [comment]
@nolouch PTAL

File: server/balancer.go [hunk 6]
```diff
@@ -31,6 +30,13 @@ const (
 	bootstrapBalanceDiff  = 2
 )
 
+type BalanceType int
```
Comment (nolouch) [inline server/balancer.go]: please fix this remind by `make dev`. and then pass the ci

**Comment by disksing** [comment]
# github.com/pingcap/pd/server
server/coordinator.go:186: cannot range over status (type *StoreHotRegionInfos)
make: *** [build] Error 2

@zhangjinpeng1987 Please merge master and fix the compile error.

**Comment by disksing** [comment]
LGTM.

**Pull request merged**